### ISSUE TITLE
IIS Express Problem - app returned http status code 400 from subscription endpoint

### ISSUE THREAD

**Issue by fbridger** [dapr/dotnet-sdk#743]
## Expected Behavior
When running and debugging the ControllerSample application using IIS Express I would expect to:
- Same behavior as running the application using Kestrel
- Event Subscriptions working correctly

## Actual Behavior

Dapr is showing an error  when trying to obtain events from the application when running in IIS Express

Generated Log
´´´shell
time="2021-08-28T20:52:49.3743365-03:00" level=error msg="app returned http status code 400 from subscription endpoint" app_id=controller instance=DESKTOP-B3VIRPH scope=dapr.runtime type=log ver=1.3.0
time="2021-08-28T20:52:49.3748689-03:00" level=debug msg="app responded with subscriptions []" app_id=controller instance=DESKTOP-B3VIRPH scope=dapr.runtime type=log ver=1.3.0
´´´

## Steps to Reproduce the Problem

1. Open ControllerSample in Visual Studio 2019
2. Select IIS Express
3. Click Debug => Start Debugging
4. Execute dapr 
´´´shell
dapr run --app-id controller --app-port 5000 --log-level debug
´´´

## Proposed Fixed

- Use ´127.0.0.1´ instead of ´localhost´

Commit details:
https://github.com/fbridger/dotnet-sdk/commit/841623ebb248a554579ccf95f6170348156440db

Let me know if I should open a Pull Request

**Comment by rynowak** [comment]
@halspang - can you talk to @jkotalik about this? It doesn't seem right that we need to manually update these kinds of settings to work with both servers.

**Comment by jkotalik** [comment]
I did some investigation on this with @halspang.

Currently, when IISExpress starts in Visual studio, by default, it doesn't explicitly bind to 127.0.0.1. You can see that with the output of `netstat -ano -p tcp` where it's only listening to 0.0.0.0:PORT. By explicitly adding 127.0.0.1 in the launch settings, dapr run works.

However, this still opened up more questions. Let's now say instead of having Kestrel bind to 127.0.0.1, it binds to any port (*:PORT in launch settings). This causes `netstat -ano -p tcp` to show that the app is listening on 0.0.0.0:PORT and nothing else.

Based off the observation with IISExpress, this should also fail to be discovered by dapr. BUT, it still succeeds, meaning there is something else going on here.

The next thought was something permission based. I tried to run dapr and iisexpress in an elevated prompt on windows and they still failed to discover the address.

Finally I thought "what if I ran the kestrel as elevated"? Welp, the app still works and the app port is discovered.

My current hypothesis is that it may have something to do with HTTP.sys permissions and/or the underlying [zeroconf](https://github.com/grandcat/zeroconf/blob/master/examples/register/server.go) that is being used. When running netstat, all of the bindings are owned by HTTP.sys, which may not be discovered by zeroconf.

Also, with respect to my unfamiliarity with dapr, how does dapr announce the app port that can be bound to? Ex I start the ControllerSample, is there something in that app that is announcing the app can be discovered at some port? I'd be curious if the app isn't quite setting something correctly when using IISExpress with the binding.

**Comment by fbridger** [comment]
> Also, with respect to my unfamiliarity with dapr, how does dapr announce the app port that can be bound to? Ex I start the ControllerSample, is there something in that app that is announcing the app can be discovered at some port? I'd be curious if the app isn't quite setting something correctly when using IISExpress with the binding.

Dapr Sidecar receives the **Application Port** that is calling as an argument `--app-port 5000`
```sh
dapr run --app-id ControllerSample --app-port 5000
```

@jkotalik Is this what you were asking?

**Comment by jkotalik** [comment]
@fbridger yeah that's what I realized now. The app that is starting (the dotnet app) does nothing to indicate that the app port is 5000 and is based off the argument.

I think the fix you proposed could be fine. Curious about your use case, are you planning to use dapr and IIS purely for local development or in production as well?

**Comment by fbridger** [comment]
Only for local development. In production it will use Kestrel running the app in a Linux container.

**Comment by fbridger** [comment]
Issue fixed in PR: https://github.com/dapr/dotnet-sdk/pull/754
